iot-hub/Samples/device/PnpUsabilitySample/PnpUsabilitySample.cs

```diff
@@ -0,0 +1,211 @@
+
+            try
+            {
+                await InitializeAndOpenClientAsync();
+                await PerformOperationsAsync(cts.Token);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError($"Unrecoverable exception caught, user action is required, so exiting...: \n{ex}");
+            }
+
+            _initSemaphore.Dispose();
+        }
+
+        private async Task InitializeAndOpenClientAsync()
+        {
+            if (ShouldClientBeInitialized(s_connectionStatus))
+            {
+                // Allow a single thread to dispose and initialize the client instance.
+                await _initSemaphore.WaitAsync();
+                try
+                {
+                    if (ShouldClientBeInitialized(s_connectionStatus))
+                    {
+                        _logger.LogDebug($"Attempting to initialize the client instance, current status={s_connectionStatus}");
+
+                        // If the device client instance has been previously initialized, then dispose it.
+                        if (s_deviceClient != null)
+                        {
+                            await s_deviceClient.CloseAsync();
+                            s_deviceClient.Dispose();
+                            s_deviceClient = null;
+                        }
+                    }
+
+                    s_deviceClient = DeviceClient.CreateFromConnectionString(_deviceConnectionString, _transportType, _clientOptions);
+                    s_deviceClient.SetConnectionStatusChangesHandler(ConnectionStatusChangeHandler);
+                    _logger.LogDebug($"Initialized the client instance.");
+                }
+                finally
+                {
+                    _initSemaphore.Release();
+                }
+
+                try
+                {
+                    // Force connection now.
+                    // OpenAsync() is an idempotent call, it has the same effect if called once or multiple times on the same client.
+                    await s_deviceClient.OpenAsync();
+                    _logger.LogDebug($"Opened the client instance.");
+                }
+                catch (UnauthorizedException)
+                {
+                    // Handled by the ConnectionStatusChangeHandler
+                }
+            }
+        }
+
+        // It is not good practice to have async void methods, however, DeviceClient.SetConnectionStatusChangesHandler() event handler signature has a void return type.
+        // As a result, any operation within this block will be executed unmonitored on another thread.
+        // To prevent multi-threaded synchronization issues, the async method InitializeClientAsync being called in here first grabs a lock
+        // before attempting to initialize or dispose the device client instance.
+        private async void ConnectionStatusChangeHandler(ConnectionStatus status, ConnectionStatusChangeReason reason)
```

Comment:
I figured we can leave this code here to aid n reconnection but advise everyone to collapse this section since they don't need to work with or around it.